### PR TITLE
auto_tidyup options can be set per data type

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
`Qobj` tidyup after each mathematical operation it the option is set to `True`. This is expected when using sparse matrices, but not as useful for Dense data. This also made it a priority to implement it for new data layer while not doing any of the visible work.

This allows to set `auto_tidyup` setting per data type.

**Changelog**
Allow `auto_tidyup` settings per data type.

### CODE REVIEW

**Comment by hodgestar** [comment]
I'm wondering whether setting `auto_tidyup` per data type is a good idea? One expects the mathematical operations performed by different data backends to give the same results but setting `auto_tidyup` to `["CSR"]` by default will break that assumption quite badly. For example, matrices might become singular when sparse and not when dense.

There is also a danger of making some benchmarks artificially fast. If we're going to compare backends, we should probably look at both backends with tidyup on or off.

Is there a reason that tidyup cannot just be specialized for the new backends that need it?

File: qutip/core/qobj.py [hunk 1]
```diff
@@ -1744,6 +1752,7 @@ def iscptp(self):
     def isherm(self):
         if self._isherm is not None:
             return self._isherm
+        self.tidyup()
```
Comment (hodgestar) [inline qutip/core/qobj.py:1755]: I feel `isherm` should not call `.tidyup()` -- I would be very surprised to discover that a method called `isX` changed the underlying data of a `Qobj`.

Comment (Ericgig) [reply qutip/core/qobj.py:1755]: `_isherm_csr_full` does not work without it, but it's a bug, not a choice for faster computation. It should totally be fixed there.

**Comment by Ericgig** [comment]
`auto-tidyup` exist to remove 0s stored in sparse matrices.  `auto-tidyup` is quite messy, rounding all values under 1e-12 to 0 per default after each matrix operation can only cause problems. 

As for the benchmark, why ask for dense operations to add an extra step to increase their outputs error to the same level as sparse operations so their benchmark do not look to good in comparisons. I think it is fair that the sparse operation's timing include the time needed to manage it's structure. 

For auto-differentiation, if operations included are done with `Qobj`, it would have to differentiate tidy-up steps after each operation. But since the tidy-up operation does not have a clean derivative (`out[out < 1e-12] = 0`) it will make the auto-differentiation very inefficient.

I see `auto-tidyup` more as a necessary evil to keep sparse matrix efficient. I would argue that we should not have `auto-tidyup` in `Qobj` and have sparse matrix operation clean themselves instead.

**Comment by AGaliciaMartinez** [comment]
I would like to add that at this point CSR and Dense already _dno't_ return the same result in every operations. For example, for matmul:
```python
size = 100
random = np.random.random((size, size))
a = qt.data.Dense(random)
b = qt.data.to['CSR'](a)
np.all((a).to_array() == (b).to_array()) # True
np.all((a+a).to_array() == (b+b).to_array()) # True
#  This one is different for Dense and CSR even though the operation is the same. 
# This seems to be due to the order at which the  operations are performed with Dense
#  and CSR matmul and the fact that floating point addition is not associative. 
# (something I learned by reading the test_mathematics in qutip)
np.all((a@a).to_array() == (b@b).to_array()) # False.
```
Although the change in `auto_tidyup` will probably make bigger the difference, given that with tidy-up "all" we can not ensure the same output for every operation, I would say that it should be ok to set auto_tidyup to `[CSR]`.

**Comment by jakelishman** [comment]
Yeah, auto-tidyup should be a function of CSR matrices only.  It doesn't make sense for dense matrices to have it, that's just burning cycles.  As Asier mentioned, you already can't rely on floating-point operations being exactly identical between different matrix types, so there's no particular problem there.  The point about auto-differentiation is an excellent reason that it must be allowed to vary between types as well.

To be honest, I'm not sure how many cases the auto-tidyup operation actually helps in keeping things efficient.  It may well be worth us having a proper investigation, and seeing how often it actually helps.  I can tweak the addition and matrix-multiplication sparse algorithms to automatically tidy up when generating their output (if they don't already...), in which case we could probably entirely remove all "automatic" tidyup, and just have it be a manual operation?

`isherm_csr_full` is actually explicitly meant to handle all cases, _especially_ those involving matrices that have not been tidied up.  The logic is actually sound (it's been deployed in `master` for a while now), but looks like there's a typo made when I handled the merge up, and the data-layer is missing a few of its mathematical tests so it didn't get spotted.  I'll fix that now.

**Comment by jakelishman** [comment]
Fixed the bad merge in #1614.

**Comment by hodgestar** [comment]
I'm +1 on having auto tidy up being a feature of the individual data layers. Could the auto tidy up setting perhaps be passed down to the data layers and then each layer can handle it in a way that seems sensible to the data layer?

**Comment by Ericgig** [comment]
The consensus seems to be removing the `auto-tidyup` from `Qobj` and have it included in `CSR` operations.
I am closing this and will make a new one in that direction.

**Comment by hodgestar** [comment]
Thanks @Ericgig!

**Pull request closed without merging**